nixos/modules/services/security/acme-dns.nix

```diff
@@ -0,0 +1,412 @@
+      default = {};
+    };
+
+    configText = mkOption {
+      type = types.nullOr types.lines;
+      description = ''
+        Literal TOML configuration text. Overrides other configuration
+        options if set.
+      '';
+      default = null;
+    };
+  };
+
+  config = let
+    configFile = if cfg.configText != null
+      then pkgs.writeText "acme-dns.toml" cfg.configText
+      else let
+        baseConfig = with cfg; {
+          general = general //
+            optionalAttrs (general.nsname == null) { nsname = general.domain; };
+          inherit database;
+          # TODO: https://github.com/joohoi/acme-dns/issues/218
+          api = api // { port = toString api.port; };
+          inherit logconfig;
+        };
+
+        fullConfig = recursiveUpdate baseConfig cfg.extraConfig;
+      in pkgs.runCommand "acme-dns.toml" {} ''
+        ${pkgs.remarshal}/bin/json2toml -o $out \
+          <<<${escapeShellArg (builtins.toJSON fullConfig)}
+      '';
+  in mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = !hasInfix "@" cfg.general.nsadmin;
+        message = ''
+          Option services.acme-dns.general.nsadmin should contain a
+          valid DNS SOA RNAME-format email address with the "@" replaced
+          with ".".
+        '';
+      }
+    ];
+
+    systemd.services.acme-dns = {
+      description = "acme-dns server";
+
+      # We use network-online.target to ensure that acme-dns can reach
+      # Let's Encrypt to renew its own HTTPS API certificate on
+      # startup. This might be unnecessary if acme-dns is robust
+      # enough to properly retry, in which case this could be removed.
+      #
+      # Note that this should probably *not* be replaced with
+      # network.target unless necessary; see
+      # https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/.
+      after = [ "network-online.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/acme-dns -c ${configFile}";
+
+        Restart = "always";
+        RestartSec = "10s";
+        StartLimitInterval = "1min";
+
+        # Set up /var/lib/acme-dns with appropriate permissions.
+        StateDirectory = "acme-dns";
+        StateDirectoryMode = "0700";
+
+        # Run in an isolated filesystem namespace.
+        # TODO: narrow down /etc further?
+        TemporaryFileSystem = "/:ro";
+        BindReadOnlyPaths = [ "/nix" "/etc" ];
+
+        # Run as a dynamically-assigned user.
+        DynamicUser = true;
+
+        # Needs CAP_NET_BIND_SERVICE for binding to privileged ports.
+        CapabilityBoundingSet = [ "CAP_NET_BIND_SERVICE" ];
+        AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" ];
+
+        # NoNewPrivileges is implied by DynamicUser.
+
+        # ProtectSystem=strict is implied by DynamicUser.
+
+        # ProtectHome is redundant with TemporaryFileSystem.
+
+        # PrivateTmp is implied by DynamicUser.
```

Comment:
We can probably drop these comments.

Reply:
The intent here is to allow cross-referencing against the systemd documentation so that it's easier to check if any hardening option is missing (e.g. because of options being added to new systemd versions), and to ensure that options aren't incorrectly dropped when other sandboxing options that imply them have to be adjusted. I copied this pattern from another local WIP module's sandboxing options and found them helpful for ensuring the service hardening options covered everything they can and made sense together, but I can drop them if you think they're too noisy.